src/cargo/core/source/source_id.rs

```diff
@@ -335,6 +335,14 @@ impl SourceId {
         }
         self.hash(into)
     }
+
+    pub fn full_eq(&self, other: &SourceId) -> bool {
+        self.inner == other.inner
+    }
+
+    pub fn full_hash<S: hash::Hasher>(&self, into: &mut S) {
+        self.inner.hash(into)
```

Comment:
This can just hash the pointer. It won't be stable between runs, but it doesn't have to be for other interning to be correct.